### PR TITLE
Vanilla Space Contract

### PR DESCRIPTION
**Pull request by pscott**
As per what was described in [milestone1](https://github.com/snapshot-labs/sx-core/pull/42/files/56e5638697ad7360f2e76265b4d74d05a41b40b6#diff-d404420bd39a9b8f0b2fd88633cb896d9ad8c7855d74de505fa51dffa35accbb), here is a vanilla space contract (missing correct types and tests).

### CODE REVIEW

File: contracts/starknet/space/space.cairo
Comment (ShahakShama) [inline contracts/starknet/space/space.cairo]: Consider defining a function called authenticator_only that checks the caller is the authenticator. Something that looks like a modifier in solidity. We do something similar in [Starkgate](https://github.com/starkware-libs/starkgate-contracts/blob/02f61b33b5177d6d75c6ee20c61d22d3cd97de9f/src/starkware/starknet/apps/starkgate/cairo/ERC20.cairo#L107)

File: contracts/starknet/space/space.cairo
Comment (ShahakShama) [inline contracts/starknet/space/space.cairo]: I think this should map into an array of votes and not a single vote. If so, we should make the following changes
1. change the type to Vote*
2. add a storage_var called n_votes
3. call alloc in the constructor and write the return value into votes

File: contracts/starknet/space/space.cairo
Comment (ShahakShama) [inline contracts/starknet/space/space.cairo]: Put this in a different file (contracts/starknet/strategies/IVotingStrategy.cairo)

File: contracts/starknet/space/space.cairo
Comment (ShahakShama) [inline contracts/starknet/space/space.cairo]: Implement these structs in a different file so that we can use them in other contracts

File: contracts/starknet/space/space.cairo
Comment (ShahakShama) [inline contracts/starknet/space/space.cairo]: There's no need for this TODO as we don't have enum in Cairo.
What we do instead is that we create constants of the numbers of the enum, and put all these constants under the same namespace.
Therefore, we won't have a choice type and this will remain felt

File: contracts/starknet/space/space.cairo
Comment (ShahakShama) [reply contracts/starknet/space/space.cairo]: Actually, this doesn't work as the data in the Vote* itself isn't accessible
We need to have some sort of vote_id or something

File: contracts/starknet/space/space.cairo
Comment (pscott) [reply contracts/starknet/space/space.cairo]: good idea :)
[309caaa](https://github.com/snapshot-labs/sx-core/pull/47/commits/309caaa37d3b57d01e5e7b81344476e4e06a8945)

File: contracts/starknet/space/space.cairo
Comment (pscott) [reply contracts/starknet/space/space.cairo]: https://github.com/snapshot-labs/sx-core/pull/47/commits/cfedfb08beec24e956cf59f855a2118e54051567 :D

File: contracts/starknet/space/space.cairo
Comment (pscott) [reply contracts/starknet/space/space.cairo]: I was referring to your Choice implementation here:
https://github.com/snapshot-labs/sx-core/commit/8f760dac692479e52bdfc26eda77c33b5a9ace9a

I think we should use this enum here in the Voting structure. What do you think about it ? :)

Comment (pscott) [reply contracts/starknet/space/space.cairo]: Ok my bad, I thought we could use namespace as `types` but it looks like we can't :) I guess I'm just used to Rust and C :)

Fix: [Remove comments asking to use Choice as a type (impossible)](https://github.com/snapshot-labs/sx-core/pull/47/commits/e120f74a7b050abc3fd24e9055a006d92f0828f2)

File: contracts/starknet/space/space.cairo
Comment (pscott) [reply contracts/starknet/space/space.cairo]: https://github.com/snapshot-labs/sx-core/pull/47/commits/cfedfb08beec24e956cf59f855a2118e54051567 :D

File: contracts/starknet/space/space.cairo
Comment (pscott) [reply contracts/starknet/space/space.cairo]: Do you know if there's a way for us to check if the `choice` argument is one of the `Choice` possible value ? Other than doing some `if .... else if ... else if...` for every possible value ? (I'd like to use it in the `vote` function, right before inserting the vote in the storage variable, to make sure the vote is valid)

**Comment by pscott** [comment]
Tests are currently not working (because state is not updating... don't know why) :) 

EDIT: fixed by using `invoke` instead of `call`...

File: contracts/starknet/space/space.cairo
Comment (pscott) [inline contracts/starknet/space/space.cairo]: We might be able to remove this bad boy if we use this: https://github.com/snapshot-labs/sx-core/pull/42#discussion_r818913667

(and simply  check that `Choice.Against <= choice <= Choice.Abstain`)

Comment (pscott) [reply contracts/starknet/space/space.cairo]: (implemented on this branch if you want to 👀 it : https://github.com/snapshot-labs/sx-core/commit/7b2060fe034945aa98ea351fe21f4cce7ed229bb)

File: contracts/starknet/objects/proposal.cairo
Comment (bonustrack) [inline contracts/starknet/objects/proposal.cairo]: Did you tried with timestamp here?

File: test/starknet/vanilla_space.ts
Comment (bonustrack) [inline test/starknet/vanilla_space.ts]: This works too
```suggestion
      const { proposal_info } = await vanillaSpace.call('get_proposal_info', {
```

File: contracts/starknet/space/space.cairo
Comment (bonustrack) [inline contracts/starknet/space/space.cairo:193]: Can you emit event here with all the info + proposer_address

File: contracts/starknet/space/space.cairo
Comment (bonustrack) [reply contracts/starknet/space/space.cairo]: If it work i'm all for it, is there a way to compare the number of steps and amount of storage required?

File: contracts/starknet/space/space.cairo
Comment (bonustrack) [inline contracts/starknet/space/space.cairo:144]: Can you also emit event here?

File: test/starknet/shared/types.ts
Comment (bonustrack) [inline test/starknet/shared/types.ts]: If we can't use 0 for "Against", it's better to have the choice in the order it will be render on UI like this:
1: For
2: Against
3: Abstain

File: contracts/starknet/space/space.cairo
Comment (bonustrack) [inline contracts/starknet/space/space.cairo]: If it's just for UI we may not need these views functions, we would just need to emit events with the space settings and the API would store this information.

File: test/starknet/shared/types.ts
Comment (pscott) [reply test/starknet/shared/types.ts]: Ok! Was following the order of `choice.cairo`. I've updated everything at the same time in https://github.com/snapshot-labs/sx-core/pull/47/commits/355c2f92425bdcdc43a4d6d6b76b91e8b485e373 :)

File: contracts/starknet/space/space.cairo
Comment (pscott) [reply contracts/starknet/space/space.cairo]: Yeah I initially wrote those with testing in mind. I've removed everything except `get_vote_info` and `get_proposal_info` in [2a1fde7](https://github.com/snapshot-labs/sx-core/pull/47/commits/2a1fde787effff41c2b618dd0c058d31e57020c4) :)

File: contracts/starknet/space/space.cairo
Comment (pscott) [reply contracts/starknet/space/space.cairo:144]: [5997c7b](https://github.com/snapshot-labs/sx-core/pull/47/commits/5997c7b45cce5cac00f09c5bd84fe6fc4b142120) :D

File: contracts/starknet/space/space.cairo
Comment (pscott) [reply contracts/starknet/space/space.cairo:193]: [5997c7b](https://github.com/snapshot-labs/sx-core/pull/47/commits/5997c7b45cce5cac00f09c5bd84fe6fc4b142120) :D

File: contracts/starknet/objects/proposal.cairo
Comment (pscott) [reply contracts/starknet/objects/proposal.cairo]: https://github.com/snapshot-labs/sx-core/pull/47/commits/ce41532f7ea1496460fdde71f80cb5e24d558a25

File: test/starknet/vanilla_space.ts
Comment (pscott) [inline test/starknet/vanilla_space.ts:85]: I introduced `fromObj` because `proposal_info.power_for` is of type `{low: bigint; high: bigint}` when I would like a `SplitUin256{low: bigint, high: bigint}`. I tried casting by doing `const _for = (proposal_info.power_for as SplitUint256).toUint();` but didn't manage to make it work. Maybe the `.ts` god @bonustrack could help on that :D

File: contracts/starknet/space/space.cairo
Comment (pscott) [reply contracts/starknet/space/space.cairo]: I get the exact same number: `518` steps :)

Comment (pscott) [reply contracts/starknet/space/space.cairo]: I get the exact same number: `518` steps :)

File: contracts/starknet/objects/vote.cairo
Comment (Orland0x) [inline contracts/starknet/objects/vote.cairo]: We should consolidate the types either into the types.cairo file or the objects directory. I wonder if there is more overhead doing it one way rather than the other.

File: contracts/starknet/space/space.cairo
Comment (Orland0x) [inline contracts/starknet/space/space.cairo]: We should pass these params as an argument for `vote`, so that the interface doesnt change once parameters are needed. Will just be an empty array for milestone 1.

File: contracts/starknet/space/space.cairo
Comment (Orland0x) [inline contracts/starknet/space/space.cairo:124]: This will effectively set the snapshot at `current_timestamp`, which will be different for each vote. Dont we want to fix it?

Comment (bonustrack) [reply contracts/starknet/space/space.cairo:124]: Good point, this should be the proposal "start" instead

File: contracts/starknet/space/space.cairo
Comment (Orland0x) [inline contracts/starknet/space/space.cairo:134]: We should throw an assertion error if the carry is non zero. It is then up to the strategy creators to create strategies that yield voting powers that wont violate this.

File: contracts/starknet/space/space.cairo
Comment (pscott) [reply contracts/starknet/space/space.cairo]: [Use params in execute, vote, and proposal](https://github.com/snapshot-labs/sx-core/pull/47/commits/a3a55bd898bbf1061ecd8e15041e4ca7a6a6ed80) :D

File: contracts/starknet/space/space.cairo
Comment (pscott) [reply contracts/starknet/space/space.cairo:134]: [Throw if overflow happens](https://github.com/snapshot-labs/sx-core/pull/47/commits/7e8bc4562c966e2e46807f2210859a4e565158e7)

File: contracts/starknet/space/space.cairo
Comment (pscott) [reply contracts/starknet/space/space.cairo:124]: [Add ethereum block number to proposal struct](https://github.com/snapshot-labs/sx-core/pull/47/commits/b886d487560f1eea040dcb397f16e1eb08ee3e11) :D

File: contracts/starknet/objects/vote.cairo
Comment (pscott) [reply contracts/starknet/objects/vote.cairo]: @ShahakShama do you have any preference between a single `types` file and a `obj/` folder with files for each type ?

Comment (ShahakShama) [reply contracts/starknet/objects/vote.cairo]: Both are good

**Comment by pscott** [comment]
Decided to use a `lib/` folder with a single file per new struct / type. https://github.com/snapshot-labs/sx-core/pull/47/commits/8371c1e6c472ed4b00a93235f59bccce0eb4471d

Can merge as soon as I get two approvals :)

**Comment by Orland0x** [comment]
looks good

**Review by Orland0x — APPROVED** [review]

**Review by bonustrack — APPROVED** [review]
Looks great!

**Pull request merged**
